akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,643 @@
+   * broadcast elements form the ofiginal [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param clazz Type of elements this hub emits and consumes
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured.
+   */
+  def sink[T](bufferSize: Int): Sink[T, Source[T, NotUsed]] = Sink.fromGraph(new BroadcastHub[T](bufferSize))
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMaterializedValue[SinkShape[T], Source[T, NotUsed]] {
+  require(bufferSize > 0, "Buffer size must be positive")
+  require(bufferSize < 4096, "Buffer size larger then 4095 is not allowed")
+  // TODO: Check for power of two
+
+  private val Mask = bufferSize - 1
+
+  val in: Inlet[T] = Inlet("BroadcastHub.in")
+  override val shape: SinkShape[T] = SinkShape(in)
+
+  // Half of buffer size, rounded up
+  private[this] val DemandThreshold = (bufferSize / 2) + (bufferSize % 2)
+
+  private sealed trait HubEvent
+
+  private object RegistrationPending extends HubEvent
+  private final case class UnRegister(id: Long, previousOffset: Int, finalOffset: Int) extends HubEvent
+  private final case class Advance(id: Long, previousOffset: Int) extends HubEvent
+  private final case class NeedWakeup(id: Long, previousOffset: Int, currentOffset: Int) extends HubEvent
+
+  private final case class Consumer(id: Long, callback: AsyncCallback[ConsumerEvent])
+
+  private object Completed
+
+  private sealed trait HubState
+  private case class Open(callbackFuture: Future[AsyncCallback[HubEvent]], registrations: List[Consumer]) extends HubState
+  private case class Closed(failure: Option[Throwable]) extends HubState
+
+  private class BroadcastSinkLogic(_shape: Shape)
+    extends GraphStageLogic(_shape) with InHandler {
+
+    private[this] val callbackPromise: Promise[AsyncCallback[HubEvent]] = Promise()
+    private[this] val noRegistrationsState = Open(callbackPromise.future, Nil)
+    val state = new AtomicReference[HubState](noRegistrationsState)
+
+    @volatile private[this] var tail = 0
+    private[this] var head = 0
```

Comment:
Since I capped buffer size at 4096, using a double length wheel is probably not a bad idea. Gets rid of modulo and overflow issues in one go, at the cost of a half-empty wheel...